cms/private_media/signal_handlers.py

```diff
@@ -0,0 +1,89 @@
+from typing import TYPE_CHECKING, Any
+
+from django.contrib.contenttypes.models import ContentType
+from django.core.exceptions import ImproperlyConfigured
+from django.db.models import IntegerField
+from django.db.models.functions import Cast
+from wagtail.models import Page, ReferenceIndex
+from wagtail.signals import page_published, page_unpublished, published, unpublished
+
+from cms.private_media.constants import Privacy
+from cms.private_media.models import PrivateImageMixin
+from cms.private_media.utils import get_private_media_models
+
+if TYPE_CHECKING:
+    from django.db.models import Model
+
+
+def publish_media_on_publish(instance: "Model", **kwargs: Any) -> None:
+    """Signal handler to be connected to the 'page_published' and 'published'
+    signals for all publishable models. It is responsible for identifying any
+    privacy-controlled media used by the object, and ensuring that it is also
+    made public.
+    """
+    for model_class in get_private_media_models():
+        model_ct = ContentType.objects.get_for_model(model_class)
+        referenced_pks = (
+            ReferenceIndex.get_references_for_object(instance)
+            .filter(to_content_type=model_ct)
+            .annotate(int_object_id=Cast("to_object_id", output_field=IntegerField()))
+            .values_list("int_object_id", flat=True)
+            .distinct()
+        )
+        queryset = model_class.objects.filter(pk__in=referenced_pks, _privacy=Privacy.PRIVATE)
+        if issubclass(model_class, PrivateImageMixin):
+            queryset = queryset.prefetch_related("renditions")
+
+        if hasattr(model_class.objects, "bulk_make_public"):
+            model_class.objects.bulk_make_public(queryset)
+        else:
+            raise ImproperlyConfigured(
+                f"The manager for {model_class.__name__} is missing a bulk_make_public() method implementation. "
+                "Did you override the manager class and forget to subclass PrivateMediaModelManager?",
+            )
+
+
+def unpublish_media_on_unpublish(instance: "Model", **kwargs: Any) -> None:
+    """Signal handler to be connected to the 'page_unpublished' and 'unpublished'
+    signals for all publishable models. It is responsible for identifying any
+    privacy-controlled media used solely by the object, and ensuring that it is
+    made private again.
+    """
+    page_ct = ContentType.objects.get_for_model(Page)
+    draft_page_ids = {str(pk) for pk in Page.objects.filter(live=False).values_list("id", flat=True)}
```

Comment:
From what I can tell, there's no way other to query the reference index in a way that would a eliminate the references from drafts. The benefit of doing it this way is that the result can at least be shared by all of the queries for each model. Ultimately, it's a set of short strings, and there will likely be far more live pages than drafts. If it becomes a bottleneck in future, it can be looked at then - but we'd have to be in the millions of pages for it to cause any real problems.

Reply:
Returning that many rows into memory can still be costly, time and resource wise. I don't think it would be too difficult to inline the query so it's handled as a `JOIN` instead, removing the need to load all the pages into memory.

Alternatively, could this be handled in a different way? Fetch the referenced pages (a small list), then do an additional query to find which of those page ids are draft (`Page.objects.filter(live=False, id__in=...)`). That way there's no `JOIN`, but also no need to load a potentially large number of ids into memory and over the network.

Reply:
> Alternatively, could this be handled in a different way? Fetch the referenced pages (a small list), then do an additional query to find which of those page ids are draft (Page.objects.filter(live=False, id__in=...)). That way there's no JOIN, but also no need to load a potentially large number of ids into memory and over the network.

It would result in an additional query for each model, but I suppose that's a trade-off worth making. Will revise as suggested.

Reply:
Revised in [a2748a6](https://github.com/ONSdigital/dis-wagtail/pull/46/commits/a2748a6721f1fd8f418be13ba972a798af82e6bd)